### PR TITLE
Device-capable pretty-printer

### PR DESCRIPTION
**Pull request by upsj**
This PR makes the pretty-printer device-capable by allocating a host-side buffer and copying to it on demand. Please feel free to play around with it for a bit, since I have only tested it in simple examples.

```
$ gdb ./simple-solver
...
(gdb) b 130
...
(gdb) run hip
...
(gdb) print A->values_
$5 = gko::Array<double> of length 147 on gko::HipExecutor (0x7ffe6f51c000) = {3, 1, 1, 1, 1, 
  1, 1, 5, 1, 1, 1, 1, 1, 1, 1, 7, 1, 1, 1, 1, 1, 1, 1, 11, 1, 1, 1, 1, 1, 1, 1, 13, 1, 1, 
  1, 1, 1, 1, 1, 17, 1, 1, 1, 1, 1, 1, 1, 19, 1, 1, 1, 1, 1, 1, 1, 23, 1, 1, 1, 1, 1, 1, 1, 
  1, 29, 1, 1, 1, 1, 1, 1, 1, 1, 31, 1, 1, 1, 1, 1, 1, 1, 1, 37, 1, 1, 1, 1, 1, 1, 1, 1, 41, 
  1, 1, 1, 1, 1, 1, 1, 43, 1, 1, 1, 1, 1, 1, 1, 47, 1, 1, 1, 1, 1, 1, 1, 53, 1, 1, 1, 1, 1, 
  1, 1, 59, 1, 1, 1, 1, 1, 1, 1, 61, 1, 1, 1, 1, 1, 1, 1, 67, 1, 1, 1, 1, 1, 1, 71}
```

### CODE REVIEW

**Comment by upsj** [comment]
We might be able to make this work for all device executors by calling exec->raw_copy*, I just didn't want to use too much C++-specific stuff

**Review by pratikvn — APPROVED** [review]
Works nicely for me! Thanks for adding this. Will be pretty useful.

File: dev_tools/scripts/gdb-ginkgo.py [hunk 1]
```diff
@@ -64,62 +57,108 @@ def get_unique_ptr_data_ptr(val):
...
+        self.pointer = get_unique_ptr_data_ptr(self.val['data_'])
 
     def children(self):
-        if self.is_cpu:
-            return self._iterator(self.pointer, self.val['num_elems_'])
-        return []
+        return self._iterator(self.execname, self.pointer, self.val['num_elems_'])
 
     def to_string(self):
         return ('%s of length %d on %s (%s)' % (str(self.val.type), int(self.val['num_elems_']), self.execname, self.pointer))
```
Comment (greole) [inline dev_tools/scripts/gdb-ginkgo.py:156]: This line is too long.
```suggestion
        return "%s of length %d on %s (%s)" % (
            str(self.val.type),
            int(self.val["num_elems_"]),
            self.execname,
            self.pointer,
        )
```

One could also use `.format()` here to be consistent

File: dev_tools/scripts/gdb-ginkgo.py [hunk 2]
```diff
@@ -64,62 +57,108 @@ def get_unique_ptr_data_ptr(val):
...
+        def update_buffer(self):
+            if self.buffer and self.buffer_count >= self.buffer_size:
+                self.buffer_item = gdb.parse_and_eval(
+                    hex(self.buffer)).cast(self.item.type)
+                self.buffer_count = 0
+                self.buffer_start = self.count
+                if self.exec == "gko::CudaExecutor":
+                    memcpy_expr = "(cudaError)cudaMemcpy({}, {}, {}, cudaMemcpyDeviceToHost)"
+                elif self.exec == "gko::HipExecutor":
+                    memcpy_expr = "(hipError_t)hipMemcpy({}, {}, {}, hipMemcpyDeviceToHost)"
```
Comment (greole) [inline dev_tools/scripts/gdb-ginkgo.py:110]: Overlong lines
```suggestion
                if self.exec == "gko::CudaExecutor":
                    memcpy_expr = (
                        "(cudaError)cudaMemcpy({}, {}, {}, cudaMemcpyDeviceToHost)"
                    )
                elif self.exec == "gko::HipExecutor":
                    memcpy_expr = (
                        "(hipError_t)hipMemcpy({}, {}, {}, hipMemcpyDeviceToHost)"
                    )
```

File: dev_tools/scripts/gdb-ginkgo.py [hunk 3]
```diff
@@ -64,62 +57,108 @@ def get_unique_ptr_data_ptr(val):
...
     "Print a gko::Array"
 
     class _iterator(Iterator):
-        def __init__ (self, start, size):
+        def __init__(self, exec, start, size):
+            self.exec = exec
             self.item = start
             self.size = size
             self.count = 0
+            if exec == "gko::CudaExecutor" or exec == "gko::HipExecutor":
```
Comment (greole) [inline dev_tools/scripts/gdb-ginkgo.py:88]:
```suggestion
            if exec in [ "gko::CudaExecutor", "gko::HipExecutor"]:
```

File: dev_tools/scripts/gdb-ginkgo.py [hunk 4]
```diff
@@ -64,62 +57,108 @@ def get_unique_ptr_data_ptr(val):
...
-            count = self.count
-            self.count = self.count + 1
-            if self.count > self.size:
+            if self.count >= self.size:
                 raise StopIteration
-            elt = self.item.dereference()
+            if self.buffer:
+                self.update_buffer()
+                elt = self.buffer_item.dereference()
+                self.buffer_item = self.buffer_item + 1
```
Comment (greole) [inline dev_tools/scripts/gdb-ginkgo.py:137]: I guess `+=1 ` should work here

File: dev_tools/scripts/gdb-ginkgo.py [hunk 5]
```diff
@@ -64,62 +57,108 @@ def get_unique_ptr_data_ptr(val):
...
+                self.buffer_count = self.buffer_count + 1
```
Comment (greole) [inline dev_tools/scripts/gdb-ginkgo.py:138]: I guess +=1 should work here

File: dev_tools/scripts/gdb-ginkgo.py [hunk 6]
```diff
@@ -64,62 +57,108 @@ def get_unique_ptr_data_ptr(val):
...
+            else:
+                elt = self.item.dereference()
+            count = self.count
             self.item = self.item + 1
```
Comment (greole) [inline dev_tools/scripts/gdb-ginkgo.py:142]: I guess +=1 should work here

File: dev_tools/scripts/gdb-ginkgo.py [hunk 7]
```diff
@@ -64,62 +57,108 @@ def get_unique_ptr_data_ptr(val):
...
+            self.count = self.count + 1
```
Comment (greole) [inline dev_tools/scripts/gdb-ginkgo.py:143]: I guess +=1 should work here

**Review by greole — COMMENTED** [review]
In general LGTM!

I added some minor formatting and pythonic code style remarks. 
Maybe two points to consider: 
- Should long arrays be truncated, like numpy for example does ?
- Would a pretty printer for (small) matrices make sense?

**Comment by upsj** [comment]
@greole `gdb` already takes care of truncation, which is also why they use the lazily-evaluated iterator-style access. You can control the output limit via `set print elements 1000`. Printing a matrix is a bit more complex, I will think about it for a bit, but that will definitely need a new PR. Thanks for the other comments, I will address them.

**Review by greole — APPROVED** [review]

**Comment by upsj** [comment]
Merging this directly, since it doesn't involve any change to the compiled Ginkgo

**Pull request merged**